server/src/main/java/org/elasticsearch/index/mapper/DocumentParserListener.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+
+public interface DocumentParserListener {
+    sealed interface Token permits Token.FieldName, Token.StartObject, Token.EndObject, Token.StartArray, Token.EndArray,
+        Token.StringAsCharArrayValue, Token.NullValue, Token.ValueToken {
```

Comment:
Nit: keep it short, moving `Token.*` outside `Token`, e.g. `DocumentParserListerer.FieldName` instead of `DocumentParserListener.Token.FieldName` ?

Reply:
I do, but it's a taste thing.. Up to you.